extension/js/common/composer/composer-contacts.ts

```diff
@@ -185,60 +196,59 @@ export class ComposerContacts extends ComposerComponent {
-                if (errsMode === 'gentleRecipientErrs') {
-                    gentleErrInvalidEmails += rawRecipientAddrInput;
-                    this.composer.debug(`parseRenderRecipients(${errsMode}).6-b (${email}|${rawRecipientAddrInput})`);
-                } else {
-                    // maybe there could be:
-                    // Xss.sanitizeAppend(this.S.cached('input_to').siblings('.recipients'), `<span>${Xss.escape(rawRecipientAddrInput)} ${Ui.spinner('green')}</span>`);
-                    // but it seems to work well without it, so not adding until proved needed
-                    this.composer.debug(`parseRenderRecipients(${errsMode}).6-c SKIPPING HARSH ERR? (${email}|${rawRecipientAddrInput})`);
-                }
+
+    public validateEmails = (uncheckedEmails: string[]): { valid: string[], invalid: string[] } => {
+        const valid: string[] = [];
+        const invalid: string[] = [];
+        for (const email of uncheckedEmails) {
+            const parsed = Str.parseEmail(email).email;
+            if (parsed) {
+                valid.push(parsed);
             } else {
-                this.composer.debug(`parseRenderRecipients(${errsMode}).6-c (${email})`);
-                Xss.sanitizeAppend(this.composer.S.cached('input_to').siblings('.recipients'), `<span>${Xss.escape(email)} ${Ui.spinner('green')}</span>`);
-                isRecipientAdded = true;
+                invalid.push(email);
             }
         }
-        this.composer.debug(`parseRenderRecipients(${errsMode}).7.gentleErrs(${gentleErrInvalidEmails})`);
-        this.composer.S.cached('input_to').val(gentleErrInvalidEmails);
-        this.composer.debug(`parseRenderRecipients(${errsMode}).8`);
+        return { valid, invalid };
+    }
+
+    public parseRenderRecipients = (container: JQuery<HTMLElement>, force?: boolean, uncheckedEmails?: string[]): boolean => {
+        const inputTo = container.find('#input_to');
+        uncheckedEmails = uncheckedEmails || String(inputTo.val()).split(',');
+        const validationResult = this.validateEmails(uncheckedEmails);
+        let recipientsToEvaluate: RecipientElement[] = [];
+        if (validationResult.valid.length) {
+            recipientsToEvaluate = [...this.createRecipientsElements(container, validationResult.valid)];
+        }
+        const invalidEmails = validationResult.invalid.filter(em => !!em); // remove empty strings
+        if (force && invalidEmails.length) {
+            recipientsToEvaluate = [...recipientsToEvaluate, ...this.createRecipientsElements(container, invalidEmails, true)];
+            inputTo.val('');
+        } else {
+            inputTo.val(validationResult.invalid.join(','));
+        }
+        this.evaluateRecipients(recipientsToEvaluate).catch(Catch.reportErr);
         this.composer.resizeInputTo();
-        this.composer.debug(`parseRenderRecipients(${errsMode}).9`);
```

Comment:
Please add back the debug calls. You have just refactored a bunch of important code, and if there is a bug that is later discovered by the CI, you will be really thankful that the debug logs are present.

Reply:
Oh, yes, I wanted to add back the debug calls at the end of refactoring but I forgot,
Adding